### ISSUE TITLE
False positive `significant-drop-tightening` with `tokioRwLockGuard.downgrade()`

### ISSUE THREAD

**Issue by mvolfik** [rust-lang/rust-clippy#11549]
### Summary

Playground with the repro: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=842ef76625594b0ddd713af3eff6064d

Version of rustc etc. below is from my local machine where I saw this happen in a larger codebase, in playground I reproduced this with v1.72.1 & Clippy 0.1.74 (2023-09-20 3223b0b)

### Lint Name

significant-drop-tightening

### Reproducer

I tried this code:

```rust
async fn main() {
    let lock = tokio::sync::RwLock::new(4);
    let mut guard = lock.write().await;
    *guard = 5;
    let guard = guard.downgrade();
    println!("{}", *guard);
}
```

I saw this happen:

```raw
warning: temporary with significant `Drop` can be early dropped
  --> src/main.rs:8:13
   |
6  |   async fn main() {
   |  _________________-
7  | |     let lock = tokio::sync::RwLock::new(4);
8  | |     let mut guard = lock.write().await;
   | |             ^^^^^
9  | |     *guard = 5;
10 | |     let guard = guard.downgrade();
11 | |     println!("{}", *guard);
12 | | }
   | |_- temporary `guard` is currently being dropped at the end of its contained scope
   |
   = note: this might lead to unnecessary resource contention
```

I expected to see this happen:

No warning.

### Version

```text
rustc 1.73.0-nightly (4c8bb79d9 2023-07-15)
binary: rustc
commit-hash: 4c8bb79d9f565115637cc6da739f8389e79f3a29
commit-date: 2023-07-15
host: x86_64-unknown-linux-gnu
release: 1.73.0-nightly
LLVM version: 16.0.5
```
